skada/utils.py

```diff
@@ -628,6 +725,142 @@ def jac(x):
+
+    Parameters
+    ----------
+    Q : (d,d) ndarray, float64, optional
+        Quadratic cost matrix matrix
+    c : (d,) ndarray, float64, optional
+        Linear cost vector
+    A : (n,d) ndarray, float64, optional
+        Linear inequality constraint matrix.
+    b : (n,) ndarray, float64, optional
+        Linear inequality constraint vector.
+    Aeq : (n,d) ndarray, float64, optional
+        Linear equality constraint matrix .
+    beq : (n,) ndarray, float64, optional
+        Linear equality constraint vector.
+    x0 : (d,) ndarray, float64, optional
+        Initialization. Ones by default.
+    max_iter : int, optional
+        Maximum number of iterations to perform.
+
+    Returns
+    -------
+    x: (d,) ndarray
+        Optimal solution x
+    val: float
+        optimal value of the objective (None if optimization error)
+    """
+    if c is None:
+        def func(x):
+            return (1/2) * x @ (Q @ x)
+
+        def jac(x):
+            return Q @ x
+    else:
+        def func(x):
+            return (1/2) * x @ (Q @ x) + x @ c
+
+        def jac(x):
+            return Q @ x + c
+
+    if Aeq is not None:
+        x = frank_wolfe(jac, Aeq[0], beq[0], beq[0],
+                        x0, max_iter)
+    elif A is not None:
+        if A.shape[0] > 1:
+            x = frank_wolfe(jac, A[0], -b[1], b[0],
+                            x0, max_iter)
+        else:
+            x = frank_wolfe(jac, A[0], b[0], b[0],
+                            x0, max_iter)
+    else:
+        raise ValueError("`A` or `Aeq` must be given when"
+                         " using the 'frank-wolfe' solver")
+    return x, func(x)
+
+
+def frank_wolfe(jac, c, clb=1., cub=1., x0=None, max_iter=1000):
+    r"""Frank-Wolfe algorithm for convex programming
+
+    Solve the following convex optimization problem:
+
+    .. math::
+        \min_x \quad  f(x)
+
+        clb <= \\langle x, c \\rangle <= cub
+
+        x >= 0
+
+    Parameters
+    ----------
+    jac : callable
+        The jacobian of f, return a vector of shape (d,).
+    c : (d,) ndarray, float64
+        Linear constraint vector.
+    clb : float64, optional, default=1.
+        Lower bound of the linear constraint.
+    cub : float64, optional, default=1.
+        Upper bound of the linear constraint.
+    max_iter : int, optional, default=1000
+        Maximum number of iterations to perform.
+
+    Returns
+    -------
+    x: (d,) ndarray
+        Optimal solution x
+    """
+    inv_c = 1. / c
+
+    if x0 is None:
+        x0 = inv_c * ((cub - clb) / 2) / c.shape[0]
+
+    x = np.copy(x0)
+
+    for k in range(1, max_iter+1):
+        grad = jac(x)
+        product = grad * inv_c
+        index = np.argmin(product)
+        vect = np.zeros(c.shape[0])
+        if product[index] >= 0:
+            vect[index] = inv_c[index] * clb
+        else:
+            vect[index] = inv_c[index] * cub
+        lr = 2. / (k + 1.)
+        x = (1 - lr) * x + lr * vect
```

Comment:
here you have no line search for lr. if f is quadratic you should be able to find the optimal step size at each iteration. But certainly for another PR.